sale_product_kit/models/product_template.py

```diff
@@ -0,0 +1,45 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class ProductTemplate(models.Model):
+    _inherit = 'product.template'
+
+    is_kit = fields.Boolean(string="Is Kit")
+
+    sub_product_ids = fields.Many2many(
+        comodel_name='product.product',
+        string="Sub Products",
+        help="Products that are part of this kit.",
+        domain=[
+            ('sale_ok', '=', True),
+            ('type', '!=', 'combo'),
+            ('is_kit', '!=', True),
+        ],
+    )
+
+    @api.constrains('is_kit')
+    def _check_not_sub_product(self):
+        for template in self:
+            if template.is_kit:
+                kit_products = self.env['product.template'].search([
+                    '&',
+                    ('is_kit', '=', True),
+                    ('sub_product_ids', 'in', template.product_variant_ids.ids),
+                ])
+                if kit_products:
+                    raise ValidationError(_("This product cannot become a kit since it is a sub-product of another kit."))
```

Comment:
Using a `search()` inside `for` will be very heavy I guess. `@api.constrains` are triggered before creation of each record, even if we try to create records externally using CSV import, XMLrpc etc. What do you say? :thinking: 
```suggestion
    @api.constrains('is_kit')
    def _check_not_sub_product(self):
    """Ensure kit is not a sub-product of another kit."""
    kit_records = self.filtered('is_kit')
    if kit_records:
        existing_kits = self.search([('sub_product_ids', 'in', kit_records.mapped('product_variant_ids').ids)])
        if existing_kits:
            raise ValidationError(_("A product cannot be a kit if it is already a sub-product of another kit."))

```

Reply:
Understood. Will use the inbuilt methods such as mapped, filtered, etc. as much as possible to avoid compromising the performance of code.